File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {

        if ((head == nullptr) or (head->next == nullptr) or (k==0)){
            return head;
        }
        int sz = 0;
        ListNode* cur=head;
        while (cur!=nullptr) {
            cur = cur->next;
            sz+=1;
        }
        k = k%sz;
        while (k--) {
            cur = head;
            while(cur->next->next){
                cur=cur->next; 
            }
            ListNode* tmp = cur->next;
            cur->next = nullptr;
            tmp->next = head;
            head = tmp;
        }
        return head;
    }
};

File: 61-rotate-list/rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:



        if (head is None) or (head.next is None) or (k==0):
            return head

        size = 0
        cur = head
        while (cur):
            cur=cur.next
            size+=1
        k = k % size
        
        while (k>0):
            cur = head
            while(cur.next.next):
                cur = cur.next
            tmp = cur.next
            cur.next = None
            tmp.next = head
            head = tmp
            k-=1

        return head
        